fastseq/optimizer/transformers/beam_search_optimizer.py

```diff
@@ -717,12 +719,17 @@ def _generate_beam_search(
         if do_sample is False:
             beam_scores[:, 1:] = -1e9
         beam_scores = beam_scores.view(-1)  # shape (batch_size * num_beams,)
+        beams_offset = (torch.arange(0, batch_size) * num_beams).unsqueeze(1).type_as(input_ids)
+
+        cand_size = 2 * num_beams
+        cand_offsets = torch.arange(0, cand_size).type_as(input_ids)
 
         # cache compute states
         past = (encoder_outputs, None) if encoder_outputs is not None else None
 
         # done sentences
         done = [False for _ in range(batch_size)]
+        use_reorder_cache_v2 = hasattr(self, '_reorder_cache_v2')
```

Comment:
In my understanding, the changes in `_generate_beam_search` are to remove the finished sentences from the batch during the iteration. As the batch size will be changed, the behavior of `_reorder_cache` needs to be changed as well. That's why `_reorder_cache_v2` is added. These changes got 10% ~ 40% speedup as said in the email.

Is my understanding correct? If yes, could we add some comments here like this: currently only UniLM supports the feature that removes the finished hypotheses from the batch during the generation. Other models in huggingfaces do not support it yet, which requires a new version `_reorder_cache_v2` to support the dynamic changed batch size. Once all the models support this feature, we will merge `_reorder_cache_v2` and `_reorder_cache` into one.

Reply:
Yeah, you are right. ` _reorder_cache` be reserved for other people who use the hf beam search.

Reply:
Great, please add a comment to the code. It will help developers read the code.